### ISSUE TITLE
T5：Kaggle 训练管线——LLaMA-Factory QLoRA + 合并导出

### ISSUE THREAD

**Issue by sunnyday9** [sunnyday9/COC-AI-KP-uni#41]
## Parent

#36

## What to build

**Kaggle 训练管线**：LLaMA-Factory QLoRA（HF 后端）——数据集注册直吃 messages+tools（零格式转换）、QLoRA 训练配置（超参起点：lr 1e-4~2e-4 / rank 8-16 / alpha=rank 或 2x / all-linear / 2-3 epoch；序列 cap ~6k、变量块瘦身：RAG top8→4、近窗 18→8、记忆 30→12）→ `export` 合并权重。一次 Kaggle 运行（T4×2/P100 16G，30h/周额度内）产出可下载的合并模型与训练曲线。

## Acceptance criteria

- [ ] Kaggle 一次运行完成训练 + 合并导出，显存不 OOM、时长在免费额度内
- [ ] 合并权重可被 transformers/vLLM 加载
- [ ] 训练/eval loss 曲线与完整配置落盘（可复现）
- [ ] 本地一条命令准备好可上传 Kaggle 的数据与配置

## Blocked by

- #40（需要训练数据）

**Comment by sunnyday9** [comment]
**本票 agent 交付已完工**（commit `773ad45`，本地分支，未推送/未关票——关票条件 = 你的 Kaggle 人工运行验证）。

## 交付清单（验收项 4：本地一条命令）

```bash
KP_KAGGLE_USERNAME=<用户名> npm run kaggle:pack   # workspace: training
```

产物 `training/out/kaggle/coc-kp-sft-v1/`（9 文件，manifest 含全量 sha256 与对账计数）：三份 JSONL（train 1887 / eval 537 / anchors 42 与 datacard 逐条对账）+ dataset_info.json + dataset-metadata.json + QLoRA train/merge YAML + notebook（11 cells：环境检查→装依赖→sha 对账→预检→DDP 训练→曲线→合并导出→transformers 验收→归档）+ manifest。全部经 TDD 契约测试锁定（training/test/kaggle.spec.ts，28 例），全量 training 套件 76/76 绿，真实数据打包 0 违规。

## 票面与数据现实的三处冲突（均已落盘披露，非硬做）

1. **LLaMA-Factory v0.9.3 无 openai converter**（`DATASET_CONVERTERS={alpaca,sharegpt}`，PyPI 最新版）：「零格式转换」不成立——sharegpt converter 不读 `tool_calls` 字段。打包时归一化：assistant 工具调用展开为 `function` 消息、连排 tool 以 `\n\n\n` 合并（不合并会被 LF 整条静默丢弃）、连排 user 以 `\n` 合并、system 后残头 assistant 丢弃；输出消息恒为 `{role,content}`，tools 落盘为 JSON 字符串。归一化规则与 notebook 预检互相独立移植且语义一致（留档 `.scratch/lf-src/` 源码对照）。
2. **6k 序列 cap 实测严重低估**：真实 Qwen3 tokenizer 实测 train max=15740 tok（总量 ~23M/epoch）——cutoff=16384 全覆盖零截断 + liger kernel 防 16k logits OOM。
3. **P100 不可用**：算力 6.0 不支持 bitsandbytes 4-bit——仅 T4×2；Turing 无 bf16 → fp16 + 导出显式 `infer_dtype: float16`（v0.9.3 没有 `export_dtype`/`plot_loss`，写了会 unrecognized 崩）。

另有决策：anchors 三份注册但训练只挂 `coc_kp_train`（37/42 止于首次 tool_call，进训练会教「调完工具就停」且 #42 gate 偏乐观；DATA_CARD §7 授权）。以上全部在 manifest `decisions` 字段落盘（含 gpu/cutoff 两条本次评审补上的）。

## 验收项在人工侧的落地位置

- AC1（一次运行不 OOM/额度内）：notebook cell 6，时间预算 8-10.5h < 12h 会话上限
- AC2（transformers/vLLM 加载）：notebook cell 9 = transformers；vLLM 冒烟在 runbook 第 5 步补了一条一条命令
- AC3（曲线+配置落盘）：notebook 自动归档 `repro.zip`（loss 曲线/trainer_state/YAML 双份/冻结依赖/run-summary）
- 人工操作全清单：`training/data/kaggle-runbook.md`（含抽检 ≥50 开训闸——先于一切）

## 存量披露（与本票无关）

`training/eval/lib/*` 有 10 个 tsc 错误（`.ts` 后缀 import 与 messages 类型不匹配），`git stash` 验证为本票之前的存量，未在本票范围内修。

下一步：抽检 → 上传 dataset → 开 T4×2 → 跑 notebook → 回传 `run-summary.json` 后关票。想固化成交互向导可 `/wizard`。

**Comment by sunnyday9** [comment]
## v16 训练实跑结果（2026-09-07）：2 epoch 全量训练被 Kaggle 12h 会话上限终止，吞吐实测定稿

**内核**：[handemelda/coc-kp-qlora-train-t5-41](https://www.kaggle.com/code/handemelda/coc-kp-qlora-train-t5-41) v16（4-bit nf4 QLoRA · 单卡 T4 · r16/α32 · liger · grad-ckpt · cutoff 12288 · 2 epoch）
**运行**：02:49 → 14:55，12h06m 后被 Kaggle 判 CANCEL_ACKNOWLEDGED（超 12h 上限，非错误）

### 结果（产物已回传 `training/out/kaggle/runs/2026-09-07-v16-cutoff/`）

- 完成 **70/472 步（14.8%）**，loss 2.12 → **1.57 正常下降**，训练本身健康；
- **实测 575-588s/步（~9.6 分钟）** → 2 epoch 全量需 **~76 小时**——远超免费额度与 12h 会话；
- 12h 内未到 checkpoint 阈值（236 步才存），adapter 权重未落盘；
- 预检如实：cutoff 12288 下 111 条（5.9%）样本保头截断。

### 根因（离线侦查定稿）

日志实证 `Using torch SDPA`——**T4 无 flash-attn 硬件，SDPA 的 memory-efficient 注意力对 8k+ 长序列 backward 极慢**，这是 9.6 分钟/步的主因（叠加单卡）。这是配置缺陷，不是数据/代码缺陷：打包管线、QLoRA、量化、数据集全部验证通过。

### 下一步选项（供拍板，未消耗剩余额度 19.3h）

| 方案 | 做法 | 预估 |
| --- | --- | --- |
| A. cutoff 4096 + 1 epoch | 重打包 cutoff 4096（超长样本截断更激进），训 1 epoch 236 步 | ~4-8h 可完成，得到一个可进 #42 gate 的模型 |
| B. 换 flash-attn 重跑 | 开 enable_internet 装 flash-attn（T4 无硬件加速但 kernel 更快） | 加速但不可控 |
| C. 归档现状 | 接受训练侧无果，本地管线已验证，待换硬件/额度再训 | — |

**建议 A**：在免费额度现实内产出第一个可评测模型。等你拍板。

**Comment by sunnyday9** [comment]
## 侦查定稿（2026-09-07）：免费额度训不动全量 2 epoch，方案 A 亦不可行——#41 训练侧待择路

### 侦查 1：为什么慢（v16 实证定稿）
- 真实 Qwen3 tokenizer：train 全部 1887 条 ≥12k tok（p50=12280，max=15740），样本尾部几乎全是 assistant/tool（训练信号）；
- T4 无 flash-attn 硬件，SDPA 对 8k+ 长序列 backward 极慢 → **575-588s/步**，2 epoch × 472 步 = **~76h**；
- Kaggle 免费层：单会话 12h 上限 + 30h/周 → 全量 2 epoch 在免费额度内**不可行**（实证，非推测）。

### 侦查 2：为什么方案 A（cutoff 4096）不可行
- LF 截断是**保头砍尾**，而训练信号（工具链/本批/叙事）全在**尾部**；
- 真实计量：cutoff 4096/6144/8192 下 400 条抽样**全部超限**，截断即丢信号——训出来不会调工具，等于白烧额度。
（幸好重跑前先做了本地零额度侦查。）

### 训练侧结论与可选路线（等你拍板，未再消耗额度）
| 路线 | 说明 | 额度 |
| --- | --- | --- |
| 归档现状 | 管线/数据/量化全验证通过，仅吞吐不符免费额度；#41 待未来硬件/额度 | 0 |
| 分段续训 | 每会话从 checkpoint 续，多会话凑 1 epoch（~38h 拆 3-4 个 12h 会话） | ~2 周额度 |
| 降窗重训（跨票） | 改 #40 蒸馏侧瘦身（近窗/记忆/RAG 再降，样本落 4-6k，不动信号），重打包重训 | 需重跑数据生成 |

**本地可复现资产已全部落盘**：`training/out/kaggle/runs/2026-09-07-v16-cutoff/`（trainer_log/train_stdout/token-lengths/冻结依赖/YAML）、打包管线代码、notebook（v16 为可复现版本）、数据包 `handemelda/coc-kp-sft-v1`（云端 v5）。

**Comment by sunnyday9** [comment]
## system 动态段裁剪选项已实现（focus tool calling，可复现）+ 校准后可行性

### 已实现（commit a0f7ee9，80/80 测试绿）
`KP_TRIM_SYSTEM=1 npm run kaggle:pack` —— 打包时裁掉 system 的**动态注入段**（`## 记忆` / `## 故事情报` / RAG 快照），**保留全部静态 KP 规则**（防剧透/严禁事项/检定/战斗/线索/施法等工具纪律段）与**全部对话/工具链/叙事**。真实数据裁剪：train 1884/1887、eval 537/537、anchors 31/42。不动 #40 冻结源数据、不动 distill 产线、默认关（原行为不变）。

### 校准后的真实收益（真实 Qwen3 tokenizer 实测）
- 样本 **12265 → 8718 tok**（-29%）；
- **tools schema 是固定大头：5270 tok/样本**（24 工具逐字 definition，21954 字符）——裁 system 动不了它，这是降到 4096 零截断不可能的根本原因；
- 裁后 system 2466 tok + tools 5270 + 对话尾部 ~1k ≈ 8718。

### 现实可行性（额度内）
| 方案 | 步时估 | 总时 | 额度 |
| --- | --- | --- | --- |
| 瘦身 + 1 epoch（236 步） | ~4.5 min/步（O(n²) 降 2x） | ~18h | 拆 2 会话（每会话从 checkpoint 续） |
| 瘦身 + 2 epoch | ~4.5 min/步 | ~35h | 超 30h/周，不可行 |

**触发训练**（验证通过后随时可跑）：
```bash
KP_KAGGLE_USERNAME=handemelda KP_TRIM_SYSTEM=1 npm run kaggle:pack   # workspace: training
kaggle datasets version -p .../coc-kp-sft-v1 -m "trim-system v6"
kaggle kernels push -p .../run   # 会先跑 1 epoch 配置
```

**Comment by sunnyday9** [comment]
## 用户决定：放弃自训，维持 BYOK——本票关闭

经完整实跑与多轮侦查（见上），用户决定**放弃本地/Kaggle 自训模型路线**，继续使用现有 BYOK（bring-your-own-key 远端 LLM）方式。

**已执行的回退**（commit `e2acd9c` 基线，11 个训练侧 commit 已移除）：
- 云端：kernel `coc-kp-qlora-train-t5-41`、dataset `coc-kp-sft-v1`、探测 kernel 全部删除；
- 本地：`training/src/kaggle/*`、runbook、向导脚本、查看器改动全部回退；
- 保留：`.env` 与 `~/.kaggle/` Kaggle CLI 凭据（无害，可复用）、`.scratch/` 存档（含 RAG 挂起议题）。

**运行时零影响**：全程 `server/`、`shared/`、`client/` 零改动，BYOK（四协议适配器 + SSRF 门禁）完好。

**沉淀的可复用资产**（未来若改主意，全在 git 历史与 issue 评论里）：
- 调通链实证：LF v0.9.3 契约、量化枚举值 `bnb`、T4 单卡/双卡内存边界、SDPA 吞吐根因（575-588s/步）、tools schema 固定 5270 tok 等；
- system 动态段裁剪思路（focus tool calling）已实现验证，样本 12.3k→8.7k tok。

本票关闭（wontfix 性质：非缺陷，是主动放弃）。
